### ISSUE TITLE
access to scala singletons

### ISSUE THREAD

**Issue by lawremi** [rstudio/sparklyr#1384]
My understanding is that Scala singletons are referenced by a `MODULE$` static field in Java. I can't find any way to access fields via the Spark connection. So I guess support for fields would do the job. But maybe there's a smarter way.

For now I'll just use reflection methods to access the field, but it would be nice to have that built in, along with some conveniences for interfacing with Scala constructs. I could contribute this if you're interested.

**Comment by kevinykuo** [comment]
Related https://github.com/rstudio/sparklyr/issues/1181

**Comment by javierluraschi** [comment]
@lawremi Do you have a particular example I can reproduce?

I'm asking since I've access singletons using `invoke_static()`, for instance, you can modify the [sparkhello](https://github.com/javierluraschi/sparkhello) project to contain:

```scala
package SparkHello

object HelloWorld {
  var world: String = "World"

  def hello() : String = {
    "Hello, world! - From Scala"
  }
}

class HelloWorld {
  var world2: String = "World2"

  def hello2() : String = {
    "Hello, world2! - From Scala"
  }
}
```

Then rebuild, recompile jars, rebuild the project and run:

```
> sparklyr::invoke_static(sc, "SparkHello.HelloWorld", "world")
[1] "World"
> sparklyr::invoke_static(sc, "SparkHello.HelloWorld", "hello")
[1] "Hello, world! - From Scala"
> hw <- sparklyr::invoke_new(sc, "SparkHello.HelloWorld")
> invoke(hw, "world2")
[1] "World2"
> invoke(hw, "hello2")
[1] "Hello, world2! - From Scala"
```

Is the call pattern you are looking for not available in this example?

Contributions are obviously very welcome, so feel free to send PR our way!

**Comment by lawremi** [comment]
I am trying to call the `arrayToArrayList()` method on this [package-level Scala trait](https://github.com/hail-is/hail/blob/ac94266449d60128b205663972830a42a4626d2d/src/main/scala/is/hail/utils/Py4jUtils.scala). 

When I try to access it with:
```
invoke_static(sc, "is.hail.utils.package$", "arrayToArrayList", array(1)) 
```
it fails with an NPE during method invocation, presumably because the method is instance-level, and so passing `null` as the object is invalid:
```
Error: java.lang.NullPointerException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        ...
```
It would be nice if the Sparklyr RPC layer checked that a method really is static before passing `null` to it.

It's not clear to me why there is no Java-level static forwarder defined as in the case with a Scala "object". I can get it to work though by using reflection to access the `MODULE$` field of the `is.hail.utils.package$` class, and then calling `arrayToArrayList()` on that.

Any advice is appreciated.

**Comment by javierluraschi** [comment]
@lawremi Since [Py4jUtils](https://github.com/hail-is/hail/blob/ac94266449d60128b205663972830a42a4626d2d/src/main/scala/is/hail/utils/Py4jUtils.scala) is a [traits](https://docs.scala-lang.org/tour/traits.html) and is actually used by [package.scala](https://github.com/hail-is/hail/blob/449556d7123a6d0978926d70a32e86817dfd2906/src/main/scala/is/hail/utils/package.scala), I would expect this to work:

```
invoke_static(sc, "is.hail.utils", "arrayToArrayList", array(1))
```

If that's not the case, could you share the whole stack trace? Can you share the `spark_connect()` statement you are using or something that can help me reproduce this? I haven't tried running HAIL on Spark (which sounds really interesting BTW!) but I bet there are a couple caveats that you might have already figured out.

**Comment by lawremi** [comment]
It does not work, and fails with: 
```
Error: java.lang.ClassNotFoundException: is.hail.utils
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:264)
	at sparklyr.StreamHandler.handleMethodCall(stream.scala:100)
	at sparklyr.StreamHandler.read(stream.scala:62)
	at sparklyr.BackendHandler.channelRead0(handler.scala:51)
	at sparklyr.BackendHandler.channelRead0(handler.scala:4)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:343)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:336)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:343)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:336)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:293)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:267)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:343)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:336)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1294)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:343)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:911)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:131)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:643)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:566)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:480)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:442)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:131)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)
	at java.lang.Thread.run(Thread.java:748)
```
My non-expert understanding is that the class of the Scala package object is mapped to `is.hail.utils.package$`, and the instance is mapped to the `MODULE$` static field. See [this](https://stackoverflow.com/questions/26516361/using-vals-from-scala-package-object-in-java), for example.

It would be great to have utilities inside Sparklyr for working at the Scala level and not worrying about these weird Java-Scala mappings.

**Comment by javierluraschi** [comment]
@lawremi thanks! I honestly wasn't aware of Scala's [package objects](http://www.scala-lang.org/docu/files/packageobjects/packageobjects.html) type, I'm working on a PR to add support through `invoke_static()` for this...

**Comment by javierluraschi** [comment]
@lawremi this should unblock the static call once the PR gets merged: https://github.com/rstudio/sparklyr/pull/1387

**Comment by lawremi** [comment]
Wow, awesome, thanks for fast fixes and responsive support.

Btw, this was obviously motivated by a Hail interface for R/Bioconductor. Hopefully I'll get it to a state where I can open-source it soon.

**Comment by javierluraschi** [comment]
@lawremi We would be so excited to take a look and try it out! If there is something blocking your work, feel free to ping me in [gitter.im/rstudio/sparklyr](https://gitter.im/rstudio/sparklyr) or mail.
